File: src/main/java/org/jcodec/containers/mp4/boxes/UrlBox.java
package org.jcodec.containers.mp4.boxes;

import org.jcodec.common.io.NIOUtils;
import org.jcodec.platform.Platform;

import java.nio.ByteBuffer;

/**
 * This class is part of JCodec ( www.jcodec.org ) This software is distributed
 * under FreeBSD License
 * 
 * @author The JCodec project
 * 
 */
public class UrlBox extends FullBox {

    private String url;

    public static String fourcc() {
        return "url ";
    }

    public static UrlBox createUrlBox(String url) {
        UrlBox urlBox = new UrlBox(new Header(fourcc()));
        urlBox.url = url;
        return urlBox;
    }

    public UrlBox(Header atom) {
        super(atom);
    }

    @Override
    public void parse(ByteBuffer input) {
        super.parse(input);
        if ((flags & 0x1) != 0)
            return;
        url = NIOUtils.readNullTermStringCharset(input, Platform.UTF_8);
    }

    @Override
    protected void doWrite(ByteBuffer out) {
        super.doWrite(out);

        if (url != null) {
            NIOUtils.write(out, ByteBuffer.wrap(Platform.getBytesForCharset(url, Platform.UTF_8)));
            out.put((byte) 0);
        }
    }
    
    @Override
    public int estimateSize() {
        int sz = 13;

        if (url != null) {
            sz += Platform.getBytesForCharset(url, Platform.UTF_8).length;
        }
        return sz;
    }

    public String getUrl() {
        return url;
    }
}

File: src/main/java/org/jcodec/containers/mp4/boxes/UrlBox.kt
package org.jcodec.containers.mp4.boxes

import org.jcodec.common.io.NIOUtils
import org.jcodec.platform.Platform
import java.nio.ByteBuffer

/**
 * This class is part of JCodec ( www.jcodec.org ) This software is distributed
 * under FreeBSD License
 *
 * @author The JCodec project
 */
class UrlBox(atom: Header) : FullBox(atom) {
    var url: String? = null
        private set

    override fun parse(input: ByteBuffer) {
        super.parse(input)
        if (flags and 0x1 != 0) return
        url = NIOUtils.readNullTermStringCharset(input, Platform.UTF_8)
    }

    override fun doWrite(out: ByteBuffer) {
        super.doWrite(out)
        if (url != null) {
            NIOUtils.write(out, ByteBuffer.wrap(Platform.getBytesForCharset(url, Platform.UTF_8)))
            out.put(0.toByte())
        }
    }

    override fun estimateSize(): Int {
        var sz = 13
        if (url != null) {
            sz += Platform.getBytesForCharset(url, Platform.UTF_8).size
        }
        return sz
    }

    companion object {
        @JvmStatic
        fun fourcc(): String {
            return "url "
        }

        @JvmStatic
        fun createUrlBox(url: String?): UrlBox {
            val urlBox = UrlBox(Header(fourcc()))
            urlBox.url = url
            return urlBox
        }
    }
}